### DISCUSSION TITLE
Resetting token consumptions everyday and maintaining history

### DISCUSSION THREAD

**Discussion by haiderali22** [vladimir-bukhtoyarov/bucket4j#182] [Q&A] [answered]
Hello, I am using this library with Rest API to restrict api calls against token, e.g.  client can call api max 1000  times daily for this 
 i am using Example 1(https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/doc-pages/jcache-usage.md)  and setting token instead of IP . A scheduler is executed at 12 am everyday which should reset it values and save the current state of bucket against token in database for statistics.   For resetting it is better to reset the the bucket manually or let library do it automatically. second for stats i can user listener and save the event to database and i the end of day compute statistics or there is a better way?

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Hello @haiderali22 

> For resetting it is better to reset the bucket manually or let library do it automatically.

Currently, there is no **reset** method in the API, so the library can not do for you. The easiest way to reset will be just removing cache entries, then the bucket state will be recreated when will be detected that cache entry is missed.

> second for stats i can user listener and save the event to database and i the end of day compute statistics or there is a better way?

If you can not afford to miss consumption events, then it will be the best way, but I suppose one optimization - just use counters with a primary key(user, day) instead of storing each consumption event independently.
